c2corg_api/views/sso.py

```diff
@@ -0,0 +1,282 @@
+                               validator=colander.OneOf(default_langs),
+                               missing=None)
+    groups = colander.SchemaNode(colander.String(),
+                                 missing=None)
+
+sso_sync_schema = SsoSyncSchema()
+
+
+def sso_sync_validator(request, **kwargs):
+    if 'sso_key' not in request.json:
+        log.warning('Attempt to use sso_sync without key from {}'
+                    .format(request.client_addr))
+        request.errors.status = 403
+        return
+
+    sso_key = DBSession.query(SsoKey). \
+        filter(SsoKey.key == request.json['sso_key']). \
+        one_or_none()
+    if sso_key is None:
+        log.warning('Attempt to use sso_sync with bad key from {}'
+                    .format(request.client_addr))
+        request.errors.status = 403
+        request.errors.add('body', 'sso_key', 'Invalid')
+        return
+
+    user = None
+
+    # search user by external_id
+    sso_external_id = DBSession.query(SsoExternalId). \
+        filter(SsoExternalId.domain == sso_key.domain). \
+        filter(SsoExternalId.external_id == request.json['external_id']). \
+        one_or_none()
+    if sso_external_id is not None:
+        user = sso_external_id.user
+
+    if user is None:
+        # search user by email
+        if 'email' not in request.json:
+            request.errors.add('body', 'email', 'Required')
+            return
+        user = DBSession.query(User). \
+            filter(User.email == request.json['email']). \
+            one_or_none()
+
+    if user is None:
+        if request.json['username'] == '':
+            request.errors.add('body', 'username', 'Required')
+        if 'name' not in request.json:
+            request.json['name'] = request.json['username']
+        if 'forum_username' not in request.json:
+            request.json['forum_username'] = request.json['username']
+        if 'lang' not in request.json:
+            request.errors.add('body', 'lang', 'Required')
+        validate_unique_attribute('email', request, **kwargs)
+        validate_unique_attribute('username', request, **kwargs)
+        validate_unique_attribute(
+            'forum_username', request, lowercase=True, **kwargs)
+        validate_forum_username(request, **kwargs)
+
+    request.sso_key = sso_key
+    request.sso_external_id = sso_external_id
+    request.sso_user = user
+
+
+@resource(path='/sso_sync', cors_policy=cors_policy)
+class SsoSyncRest(object):
+    def __init__(self, request):
+        self.request = request
+
+    @json_view(
+        schema=sso_sync_schema,
+        validators=[
+            colander_body_validator,
+            sso_sync_validator,
+        ])
+    def post(self):
+        """
+        Synchronize user details and return authentication url.
+        Important: Email addresses need to be validated by external site.
+        """
+        request = self.request
+        sso_key = request.sso_key
```

Comment:
Don't you need to get the attributes from ``request.json`` or ``request.validated``?